In [1]:
# # Install a conda package in the current Jupyter kernel
# import sys
# !conda install --yes --prefix {sys.prefix} tensorflow

# All of the packages I need to do this

In [2]:
import warnings
warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

# Importing the minst fashion data and preprocessing it to be accepted into ANN

In [3]:
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

input_dim = 784  # 28*28
output_dim = nb_classes = 10
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255
Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

# Now we are going to do something kind of like king of the court, but king of the neural network. The one I deem the best stays and the one that isn't goes. I will go down the list of hyperparameters to tune:
 - Number of layers
 - Activation functions
 - Number of Neurons
 - Batch sizes during training
 
 
# We will see which model reigns supreme!

In [4]:
X_train.shape

(60000, 784)

In [5]:
epochs = 10
batch_size = 10

In [16]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, y_train, batch_size=batch_size, validation_data=(X_test, y_test), epochs=epochs, verbose=1)

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 4s 71us/sample - loss: 0.5967 - acc: 0.7920 - val_loss: 0.4631 - val_acc: 0.8352
Epoch 2/10
60000/60000 [==============================] - 4s 68us/sample - loss: 0.4269 - acc: 0.8492 - val_loss: 0.4482 - val_acc: 0.8390
Epoch 3/10
60000/60000 [==============================] - 4s 70us/sample - loss: 0.3853 - acc: 0.8615 - val_loss: 0.4106 - val_acc: 0.8519
Epoch 4/10
60000/60000 [==============================] - 4s 74us/sample - loss: 0.3575 - acc: 0.8701 - val_loss: 0.4041 - val_acc: 0.8567
Epoch 5/10
60000/60000 [==============================] - 4s 71us/sample - loss: 0.3383 - acc: 0.8780 - val_loss: 0.3715 - val_acc: 0.8659
Epoch 6/10
60000/60000 [==============================] - 4s 71us/sample - loss: 0.3217 - acc: 0.8836 - val_loss: 0.3660 - val_acc: 0.8698
Epoch 7/10
60000/60000 [==============================] - 4s 69us/sample - loss: 0.3111 - acc: 0.8860 - val_loss: 0.

In [17]:
score = model.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.3742851203918457
Test accuracy: 0.8603


# I've seen worse

In [23]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# our third layer
model.add(Dense(64, activation="relu"))
# our fourth layer
model.add(Dense(32, activation="relu"))
# our fifth layer
model.add(Dense(32, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, y_train, batch_size=batch_size, validation_data=(X_test, y_test), epochs=epochs, verbose=1)

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 5s 86us/sample - loss: 0.6861 - acc: 0.7492 - val_loss: 0.4946 - val_acc: 0.8214
Epoch 2/10
60000/60000 [==============================] - 5s 88us/sample - loss: 0.4245 - acc: 0.8462 - val_loss: 0.4864 - val_acc: 0.8246
Epoch 3/10
60000/60000 [==============================] - 6s 97us/sample - loss: 0.3785 - acc: 0.8610 - val_loss: 0.4270 - val_acc: 0.8483
Epoch 4/10
60000/60000 [==============================] - 6s 92us/sample - loss: 0.3500 - acc: 0.8705 - val_loss: 0.4049 - val_acc: 0.8547
Epoch 5/10
60000/60000 [==============================] - 5s 87us/sample - loss: 0.3310 - acc: 0.8782 - val_loss: 0.3864 - val_acc: 0.8608
Epoch 6/10
60000/60000 [==============================] - 5s 86us/sample - loss: 0.3133 - acc: 0.8829 - val_loss: 0.3608 - val_acc: 0.8705
Epoch 7/10
60000/60000 [==============================] - 5s 85us/sample - loss: 0.3036 - acc: 0.8862 - val_loss: 0.

In [24]:
score = model.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.35084240040779113
Test accuracy: 0.8734


# The accuracy was a little better with this model.Model 2 moves on.


# Next we compare activation functions

In [10]:
# #Repeat model stats
# Test score: 0.34816762804985046
# Test accuracy: 0.8732
# loss: 0.3482

In [36]:
#New model
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="selu"))
# our second dense layer
model.add(Dense(64, activation="selu"))
# our third layer
model.add(Dense(64, activation="selu"))
# our fourth layer
model.add(Dense(32, activation="selu"))
# our fifth layer
model.add(Dense(32, activation="selu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, y_train, batch_size=batch_size, validation_data=(X_test, y_test), epochs=epochs, verbose=1)

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 9s 156us/sample - loss: 0.5300 - acc: 0.8105 - val_loss: 0.5502 - val_acc: 0.8205
Epoch 2/10
60000/60000 [==============================] - 9s 147us/sample - loss: 0.4260 - acc: 0.8504 - val_loss: 0.4269 - val_acc: 0.8434
Epoch 3/10
60000/60000 [==============================] - 9s 146us/sample - loss: 0.3973 - acc: 0.8617 - val_loss: 0.4658 - val_acc: 0.8411
Epoch 4/10
60000/60000 [==============================] - 9s 149us/sample - loss: 0.3845 - acc: 0.8676 - val_loss: 0.4974 - val_acc: 0.8410
Epoch 5/10
60000/60000 [==============================] - 9s 145us/sample - loss: 0.3837 - acc: 0.8709 - val_loss: 0.4667 - val_acc: 0.8562
Epoch 6/10
60000/60000 [==============================] - 9s 142us/sample - loss: 0.3923 - acc: 0.8717 - val_loss: 0.4706 - val_acc: 0.8543
Epoch 7/10
60000/60000 [==============================] - 9s 146us/sample - loss: 0.3979 - acc: 0.8714 - val_l

In [37]:
score = model.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.597858249270916
Test accuracy: 0.8288


# 2nd model still coming out on top! Accuracy difference!

In [33]:
#Number of neurons time
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(128, activation="relu"))
# our third layer
model.add(Dense(64, activation="relu"))
# our fourth layer
model.add(Dense(64, activation="relu"))
# our fifth layer
model.add(Dense(32, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, y_train, batch_size=batch_size, validation_data=(X_test, y_test), epochs=epochs, verbose=1)

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 6s 92us/sample - loss: 0.6594 - acc: 0.7667 - val_loss: 0.5690 - val_acc: 0.7984
Epoch 2/10
60000/60000 [==============================] - 6s 104us/sample - loss: 0.4172 - acc: 0.8483 - val_loss: 0.4226 - val_acc: 0.8421
Epoch 3/10
60000/60000 [==============================] - 6s 106us/sample - loss: 0.3710 - acc: 0.8658 - val_loss: 0.3998 - val_acc: 0.8556
Epoch 4/10
60000/60000 [==============================] - 6s 104us/sample - loss: 0.3440 - acc: 0.8727 - val_loss: 0.4110 - val_acc: 0.8519
Epoch 5/10
60000/60000 [==============================] - 7s 109us/sample - loss: 0.3246 - acc: 0.8801 - val_loss: 0.3678 - val_acc: 0.8643
Epoch 6/10
60000/60000 [==============================] - 6s 97us/sample - loss: 0.3094 - acc: 0.8843 - val_loss: 0.3752 - val_acc: 0.8605
Epoch 7/10
60000/60000 [==============================] - 6s 97us/sample - loss: 0.2945 - acc: 0.8899 - val_loss

In [35]:
score = model.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.35251856787204744
Test accuracy: 0.8755


# Well the accuracy is better but test score is worse. Keeping original model

In [38]:
batch_size = 15

In [ ]:
# #Repeat model stats
# Test score: 0.34816762804985046
# Test accuracy: 0.8732
# loss: 0.3482

In [39]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# our third layer
model.add(Dense(64, activation="relu"))
# our fourth layer
model.add(Dense(32, activation="relu"))
# our fifth layer
model.add(Dense(32, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, y_train, batch_size=batch_size, validation_data=(X_test, y_test), epochs=epochs, verbose=1)

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 4s 71us/sample - loss: 0.7005 - acc: 0.7455 - val_loss: 0.6746 - val_acc: 0.7462
Epoch 2/10
60000/60000 [==============================] - 4s 64us/sample - loss: 0.4409 - acc: 0.8425 - val_loss: 0.4456 - val_acc: 0.8410
Epoch 3/10
60000/60000 [==============================] - 4s 65us/sample - loss: 0.3891 - acc: 0.8591 - val_loss: 0.4196 - val_acc: 0.8464
Epoch 4/10
60000/60000 [==============================] - 4s 63us/sample - loss: 0.3597 - acc: 0.8674 - val_loss: 0.3955 - val_acc: 0.8581
Epoch 5/10
60000/60000 [==============================] - 4s 64us/sample - loss: 0.3386 - acc: 0.8751 - val_loss: 0.3904 - val_acc: 0.8593
Epoch 6/10
60000/60000 [==============================] - 4s 64us/sample - loss: 0.3232 - acc: 0.8799 - val_loss: 0.3697 - val_acc: 0.8664
Epoch 7/10
60000/60000 [==============================] - 4s 66us/sample - loss: 0.3107 - acc: 0.8853 - val_loss: 0.

In [40]:
score = model.evaluate(X_test, y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.3646376796722412
Test accuracy: 0.8725


# That model again! Let's look at them all together!

 - Model 1: Test score: 0.3742851203918457, Test accuracy: 0.8603
 - Model 2: Test score: 0.35084240040779113, Test accuracy: 0.8734 
 - Model 3: Test score: 0.35251856787204744, Test accuracy: 0.8755
 - Model 4: Test score: 0.3646376796722412, Test accuracy: 0.8725

# Model 3 might have performed the best out of all the models. I think tuning of the epochs as wel as neurons could lead to increased accuracy and overall better test score.